README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/alisomay/smrec/main/assets/logo_transparent.png"/>
</p>

# smrec

Minimalist multi-track audio recorder which may be controlled via OSC or MIDI.

I did this because I needed a simple multi-track audio recorder which I could control via OSC or MIDI.

I didn't want and do not have the resources to use a DAW for this purpose in my setup and wanted mono wave files for each channel organized in a directory per recording by date and time

I'm using this recorder in a setup where I use a [Behringer XR18](https://www.behringer.com/product.html?modelCode=P0BI8) as an audio interface and a [LattePanda 3 Delta](https://www.lattepanda.com/lattepanda-3-delta) as a SBC.

Now let's record some sound! 🔔

---

## Installation

```
cargo install smrec
```

### Installing for Windows and ASIO support

`smrec` uses [`cpal`](https://github.com/RustAudio/cpal) as the underlying audio API. `cpal` supports WASAPI, DirectSound and ASIO on Windows. However, since `cargo` builds binaries from source in the target machine and it is [not very straight forward](https://github.com/RustAudio/cpal#asio-on-windows) to build `cpal` with ASIO support due to `asio-sys` build script, **there is a pre-build script provided** in this repository.

To install `smrec` on Windows, please follow these steps in order:

- Install the latest Visual Studio (if you don't have it already)
- Install the latest LLVM (if you don't have it already) from [here](https://releases.llvm.org/download.html)
- Open Command Prompt as administrator
- Set LLVM path in the environment variables (system wide)
  ```
  setx /M LIBCLANG_PATH "C:\Program Files\LLVM\bin"
  ```
- Open PowerShell as administrator
- Check if the environment variable is set correctly
  ```
  $env:LIBCLANG_PATH
  ```
- Source the pre-build script in the repository (assuming you are in the root of the repository)
  ```
  . .\pre-build-win.ps1
  ```
  this script will download the ASIO SDK, set Visual Studio environment variables and `CPAL_ASIO_DIR` variable for the current shell session.
- Install as usual
  ```
  cargo install smrec
  ```

If you know what you're doing feel free to skip these steps and consult the [`cpal` documentation](https://github.com/RustAudio/cpal#asio-on-windows).

### Pre-built binaries

Pre-built binaries as an alternative are available for Windows [here](https://github.com/alisomay/smrec/releases) due to the complicated process of building `cpal` with ASIO support on Windows currently.

## Tutorial

### Simply as a command

```
smrec
```

Runs with the default configuration which is:

- The default audio host
- The default audio device
- The default set of input channels
- The default directory to record which is the current working directory
- The recording goes on until `ctrl+c` is pressed and the program is interrupted.
- Creates a directory named `rec_YYYYMMDD_HHMMSS` in the current working directory and records the audio in that directory.
- The audio is recorded in `wav` format.
- The audio is recorded in the default sample rate and buffer size and sample format of the audio device.
- For every channel a separate file is created (mono) and the file name for each is `chn_XX.wav` where `XX` is the channel number.

To record for a specific duration, use the `--duration` flag and specify the duration in seconds.
The following command records for 10 seconds:

```
smrec --duration 10
```

By using the `--host` and `--device` flag , you can specify the audio host and device to use. The following command uses `MacBook Pro Microphone` as the audio device:

```
smrec --device "MacBook Pro Microphone"
```

#### Listing midi ports and audio hosts and devices

```
smrec list
```

#### Including and excluding channels from a recording

By default, all channels of the audio device are recorded. You can specify which channels to include or exclude from the recording by using the `--include` and `--exclude` flags. These flags can not be used together. The following command records only the first two channels of a 4 channel audio device:

```
smrec --include 1,2
```

And this command records all channels except the first channel of a 4 channel audio device:

```
smrec --exclude 1
```

as seen in the examples, the channel numbers start from 1 and they can be specified as a comma separated list.

#### Recording to a specific directory

By default, the recording is done in the current working directory. You can specify a directory to record to by using the `--directory` flag. The following command records to the `~/Music` directory:

```
smrec --out ~/Music
```

#### Configuring with a configuration file

`smrec` uses the cli arguments for configuration and they precede everything. However, you can configure some aspects (probably more to come) of `smrec` by using a configuration file so they replace the default configuration. The configuration file is a `toml` file and it is named `config.toml`. The configuration file is searched in the following order:

- `.smrec/config.toml` in the current working directory.
- `.smrec/config.toml` in the user home directory.
- If none of the above is found, the default configuration is used.

The configuration file can configure:

- Channel names

```toml
[channel_names]
1 = "Kick.wav"
2 = "Snare.wav"
3 = "Hi-Hat.wav"
```

- More to come..

### OSC control

`smrec` normally starts recording as soon as it is run. However it also has options for various control methods.

Running, `smrec --osc` will not start recording immediately but instead it will wait for an OSC message to start recording.
The default OSC port for receiving and sending is chosen randomly by the os and the default addresses for sending and receiving is `127.0.0.1` and `0.0.0.0`.
After running the command above, the output might look like this:

```
Will be sending OSC messages to 127.0.0.1:61213
Listening for OSC messages on 0.0.0.0:51014
```

Currently `smrec` does not support IPv6.

In the default configuration:

- Listens for OSC messages on a randomly chosen port on all addresses.
- Sends OSC messages to localhost on a randomly chosen port.

To configure OSC further arguments could be added to the flag:

```
smrec --osc "<listen_address>:<listen_port>;<send_address>:<send_port>"
```

or

```
smrec --osc "<listen_address>:<listen_port>"
```

the second form would keep the default send address and port.

```
smrec --osc "0.0.0.0:18000;255.255.255.255:18001"
```

will listen for OSC messages on all addresses on port `18000` and send OSC messages to all addresses on port `18001`.
Yes, `smrec` can also broadcast OSC messages is the OS and the network allows it.

#### OSC messages

The messages which `smrec` listens for are:

- `/smrec/start` - Starts the recording, sending a second start will stop the running recording and starts a new one creating a new directory in the specified root.
- `/smrec/stop` - Stops the recording if there is a running one.

The messages which `smrec` sends are:

- `/smrec/start` - Sent when a new recording is started.
- `/smrec/stop` - Sent when a running recording is stopped.
- `/smrec/error <string>`- Sent when some errors occur and the error message is transferred a string in the argument.

### MIDI control

`smrec` can also be controlled via MIDI. It can even be controlled via OSC and MIDI simultaneously.
Though `smrec` is a simplistic application to serve a single purpose the MIDI communication the options it provides for configuring MIDI is extensive.

Running, `smrec --midi` will not start recording immediately but instead it will wait for a MIDI CC message to start recording.

Here is the default configuration:

- Finds all available MIDI ports and starts listening on them.
- Listens for any channel in those ports.
- Reacts to CC 16 to start the recording and CC 17 to stop the recording.
- As in OSC, sending subsequent CC 16 messages will stop the running recording and start a new one creating a new directory in the specified root.
- `smrec --midi` is synonymous with `smrec --midi "[*[(*,16,17)]]"` which will be explained below.

#### Configuration

The `--midi` flag accepts a string argument which is parsed as a configuration string which configures the input and output.
These strings are separated by a semicolon (`;`) and the first part configures the input and the second part configures the output.
Any part could be left out.

Deconstruction:

- `[..]` is a container for an input or output configuration.
- `[port name[..], ..]` a comma separated list of port names which `smrec` will connect to.
- `[port name[(..), ..], ..]` each port name should contain at least one channel/MIDI CC filter configuration.
- `(<channel number>, <cc number for starting the recording>, <cc number for stopping the recording>)` this is the structure of a channel/MIDI CC filter configuration.
- `(1,2,3)` here is an example, this will listen for CC 2 on channel 2 to start the recording and CC 3 on channel 2 to stop the recording. All other messages in that port is ignored. MIDI channels are 0 indexed!
- `[my nice port[(1,2,3), ..], ..]` this is how we use that tuple.
- `[my nice port[(1,2,3), (15, 127, 126), ..], ..]` as all the elements we can have multiples of those.
- `[ my first port[(1,2,3), (15, 127, 126), (12,4,5)], my second port[(1,2,3)] ]` here is a valid configuration string. It will listen for CC 2 on channel 2 to start the recording and CC 3 on channel 2 to stop the recording on `my first port` and listen for CC 2 on channel 2 to start the recording and CC 3 on channel 2 to stop the recording on `my second port`. All other messages in those ports are ignored.

Use of '\*' and glob patterns:

- Port names in the configuration string are treated as [glob patterns](<https://en.wikipedia.org/wiki/Glob_(programming)>).
- `*` matches any port. Which in the end means all ports.
- All valid glob patterns could be used to match port names.
- Deconstructing the default configuration string: `[*[(*,16,17)]]` now should make sense.
- Listen on all ports, do not filter by channel reacting to all MIDI CC messages which are CC 16 to start the recording and CC 17 to stop the recording.

`smrec` can also send midi messages on certain events.
If the output port is configured with a configuration, the configured CC messages will be sent on the configured port and channels on start and stop events.

#### Values

MIDI CC values are considered momentary.

Once a value `127` is received through a configured MIDI CC number the action is taken immediately.
**This is why sending bursts of MIDI CC messages is not a good idea.**
Every message would trigger a new recording if it is configured to start the recording.

`smrec` sends MIDI CC messages with a value of `127` on start and `127` on stop to the configured MIDI CC numbers if output is configured.

As a last example to get the hang of it, this configuration string will listen for CC 2 on channel 2 to start the recording and CC 3 on channel 2 to stop the recording on `my first port` and listen for CC 2 on channel 2 to start the recording and CC 3 on channel 2 to stop the recording on `my second port`. All other messages in those ports are ignored. On start and stop events, it will send CC 16 with a value of 127 on channel 2 on `my first port` and send CC 17 with a value of 127 on channel 2 on `my second port`.

```
[ my first port[(1,2,3), (15, 127, 126), (12,4,5)], my second port[(1,2,3)] ];[ my first port[(1,2,3), (15, 127, 126), (12,4,5)], my second port[(1,2,3)] ]
```

## Next steps

I'm going to make sure,

- Installation gets smoother
- Proper distribution packages are provided
- Documentation is complete
- Bugs are fixed
- Better messages to the user

But I don't plan to heavily maintain this project, I'll just make sure that it is usable enough and lives.

## Support

- Desktop
  - macOS:
    - `x86_64` ✅
    - `aarch64` ✅
  - linux:
    - `x86_64` ✅
    - `aarch64` ✅
  - windows:
    - `x86_64` ✅
    - `aarch64` ✅

## Contributing

- Be friendly and productive
- Follow common practice open source contribution culture
- Rust [code of conduct](https://www.rust-lang.org/policies/code-of-conduct) applies

Thank you 🙏

## Last words

It is something I needed to resolve a specific problem and I shared it publicly.
I hope it resolves your problem too.


CHANGELOG.md
# Changelog

The format of this changelog is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

smrec project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- ## Unreleased

### Changed

## [x.x.x] <date>

### Added

#### Summary per area

- Flat list

### Changed

#### Summary per area

- Flat list

### Removed

#### Removed thing or area

- Declaration
  - The reason
  - See https://github.com/koto-lang/koto/issues/201 for removal rationale.

### Fixed

- Topic
  - Details -->

## [0.2.1] - 2020.11.20

- Initial release


src/main.rs
// Most of the lints we deny here have a good chance to be relevant for our project.
#![deny(clippy::all)]
// We warn for all lints on the planet. Just to filter them later for customization.
// It is impossible to remember all the lints so a subtractive approach keeps us updated, in control and knowledgeable.
#![warn(clippy::pedantic, clippy::nursery, clippy::cargo)]
// Then in the end we allow ridiculous or too restrictive lints that are not relevant for our project.
// This list is dynamic and will grow in time which will define our style.
#![allow(
    clippy::multiple_crate_versions,
    clippy::blanket_clippy_restriction_lints,
    clippy::missing_docs_in_private_items,
    clippy::pub_use,
    clippy::std_instead_of_alloc,
    clippy::std_instead_of_core,
    clippy::implicit_return,
    clippy::missing_inline_in_public_items,
    clippy::similar_names,
    clippy::question_mark_used,
    clippy::expect_used,
    clippy::missing_errors_doc,
    clippy::pattern_type_mismatch,
    clippy::module_name_repetitions,
    clippy::empty_structs_with_brackets,
    clippy::as_conversions,
    clippy::self_named_module_files,
    clippy::cargo_common_metadata,
    clippy::exhaustive_structs,
    // It is a binary crate, panicing is usually fine.
    clippy::missing_panics_doc
)]

mod config;
mod list;
mod midi;
mod osc;
mod stream;
mod types;
mod wav;

use crate::{
    config::{choose_channels_to_record, SmrecConfig},
    midi::Midi,
};
use anyhow::{bail, Result};
use clap::{Parser, Subcommand};
use config::{choose_device, choose_host};
use cpal::traits::{DeviceTrait, StreamTrait};
use hound::WavWriter;
use osc::Osc;
use std::{
    cell::RefCell,
    fs::File,
    io::BufWriter,
    rc::Rc,
    sync::{Arc, Mutex},
};
use types::Action;

#[derive(Parser)]
#[command(
    author,
    version,
    about = "Minimalist multi-track audio recorder which may be controlled via OSC or MIDI.
You may visit <https://github.com/alisomay/smrec/blob/main/README.md> for a detailed tutorial."
)]
struct Cli {
    /// Specify audio host.
    /// Example: smrec --host "Asio"
    #[clap(long)]
    host: Option<String>,
    /// Specify audio device.
    /// Example: smrec --device "MacBook Pro Microphone"
    #[clap(long)]
    device: Option<String>,
    /// Include specified channels in recording.
    /// Example: smrec --include 1,2
    #[clap(long, value_delimiter = ',', num_args = 1..)]
    include: Option<Vec<usize>>,
    /// Exclude specified channels from recording.
    /// Example: smrec --exclude 1
    #[clap(long, value_delimiter = ',', num_args = 1..)]
    exclude: Option<Vec<usize>>,
    /// Specify path to configuration file.
    /// Example: smrec --config "./config.toml"
    #[clap(long)]
    config: Option<String>,
    /// Specify directory for recording output.
    /// Example: smrec --out ~/Music
    #[clap(long)]
    out: Option<String>,
    /// Specify recording duration in seconds.
    /// Example: smrec --duration 10
    #[clap(long)]
    duration: Option<String>,
    /// Configure OSC control.
    /// Example: smrec --osc "0.0.0.0:18000;255.255.255.255:18001"
    #[clap(long, value_delimiter = ';', num_args = 0..2, default_value = "EMPTY_HACK", hide_default_value = true)]
    osc: Vec<String>,
    /// Configure MIDI control.
    /// Example: smrec --midi my first port[(1,2,3), (15, 127, 126), (12,4,5)], my second port[(1,2,3)]
    #[clap(long, value_delimiter = ';', num_args = 0..2, default_value = "EMPTY_HACK", hide_default_value = true)]
    midi: Vec<String>,

    #[clap(subcommand)]
    command: Option<Commands>,
}

#[derive(Subcommand)]
enum Commands {
    /// Lists hosts, devices and configs.
    #[clap(about = "Lists hosts, devices and configs.")]
    List(List),
}

#[derive(Parser)]
struct List {
    /// List MIDI configurations.
    /// Example: smrec list --midi
    #[clap(long)]
    midi: bool,
    /// List audio configurations.
    /// Example: smrec list --audio
    #[clap(long)]
    audio: bool,
}

pub type WriterHandle = Arc<Mutex<Option<WavWriter<BufWriter<File>>>>>;
pub type WriterHandles = Arc<Vec<WriterHandle>>;

#[allow(clippy::too_many_lines)]
fn main() -> Result<()> {
    let cli = Cli::parse();

    let host = choose_host(cli.host)?;

    if let Some(command) = cli.command {
        match command {
            // Enumerate and exit.
            Commands::List(list) => {
                if list.midi {
                    list::enumerate_midi()?;
                }
                if list.audio {
                    list::enumerate_audio()?;
                }
                if !list.audio || !list.midi {
                    list::enumerate_audio()?;
                    println!();
                    list::enumerate_midi()?;
                }
            }
        };
        return Ok(());
    }

    let device = choose_device(&host, cli.device)?;
    let writers_container: Arc<Mutex<Option<WriterHandles>>> = Arc::new(Mutex::new(None));
    let stream_container: Rc<RefCell<Option<cpal::Stream>>> = Rc::new(RefCell::new(None));

    if let Ok(config) = device.default_input_config() {
        let smrec_config = Arc::new(SmrecConfig::new(
            cli.config,
            cli.out,
            choose_channels_to_record(cli.include, cli.exclude, &config)?,
            config.clone(),
        )?);

        let (to_main_thread, from_listener_thread) = crossbeam::channel::unbounded::<Action>();
        let (to_listener_thread, from_main_thread) = crossbeam::channel::unbounded::<Action>();

        let cli_osc = if cli.osc == vec!["EMPTY_HACK"] {
            None
        } else if cli.osc.is_empty() {
            Some(vec![])
        } else {
            Some(cli.osc)
        };

        let cli_midi = if cli.midi == vec!["EMPTY_HACK"] {
            None
        } else if cli.midi.is_empty() {
            Some(vec![])
        } else {
            Some(cli.midi)
        };

        let osc = if let Some(osc_config) = cli_osc {
            if osc_config.len() > 2 {
                bail!("Too many arguments for --osc");
            }
            let mut osc = Osc::new(
                &osc_config,
                to_main_thread.clone(),
                from_main_thread.clone(),
            )?;
            osc.listen();
            Some(osc)
        } else {
            None
        };

        let midi = if let Some(midi) = cli_midi {
            let mut midi = Midi::new(to_main_thread, from_main_thread, &midi)?;
            midi.listen()?;
            Some(midi)
        } else {
            None
        };

        match (midi, osc) {
            (None, None) => {
                // Pass
            }
            _ => listen_and_block_main_thread(
                &from_listener_thread,
                &to_listener_thread,
                &device,
                &stream_container,
                &writers_container,
                &smrec_config,
            ),
        }

        // No listeners, just start recording, for ever or for a certain duration.

        new_recording(
            &device,
            &stream_container,
            &writers_container,
            &smrec_config,
        )?;

        cli.duration.map_or_else(
            || {
                std::thread::park();
            },
            |dur| {
                let secs = dur
                    .parse::<u64>()
                    .expect("--duration must be a positive integer.");
                std::thread::park_timeout(std::time::Duration::from_secs(secs));
            },
        );

        stop_recording(&stream_container, &writers_container)?;
        println!("Recording complete!");
    } else {
        bail!("No default input config found for device.");
    }

    Ok(())
}

pub fn listen_and_block_main_thread(
    from_listener_thread: &crossbeam::channel::Receiver<Action>,
    to_listener_thread: &crossbeam::channel::Sender<Action>,
    device: &cpal::Device,
    stream_container: &Rc<RefCell<Option<cpal::Stream>>>,
    writers_container: &Arc<Mutex<Option<WriterHandles>>>,
    smrec_config: &SmrecConfig,
) {
    loop {
        match from_listener_thread.recv() {
            Ok(Action::Start) => {
                if let Err(err) =
                    new_recording(device, stream_container, writers_container, smrec_config)
                {
                    println!("Error starting recording: {err}");

                    to_listener_thread
                        .send(Action::Err(format!("Error starting recording: {err}")))
                        .expect("Internal thread error.");
                } else {
                    to_listener_thread
                        .send(Action::Start)
                        .expect("Internal thread error.");
                }
            }
            Ok(Action::Stop) => {
                if let Err(err) = stop_recording(stream_container, writers_container) {
                    println!("Error stopping recording: {err}");
                    to_listener_thread
                        .send(Action::Err(format!("Error starting recording: {err}")))
                        .expect("Internal thread error.");
                } else {
                    to_listener_thread
                        .send(Action::Stop)
                        .expect("Internal thread error.");
                }
            }
            // Should not be used here though, no user facing api anyway.
            Ok(Action::Err(err)) => {
                println!("Error: {err}");
            }
            Err(_) => {
                println!("Error receiving from listener thread.");
            }
        }
    }
}

pub fn new_recording(
    device: &cpal::Device,
    stream_container: &Rc<RefCell<Option<cpal::Stream>>>,
    writer_handles: &Arc<Mutex<Option<WriterHandles>>>,
    smrec_config: &SmrecConfig,
) -> Result<()> {
    // If there's an active stream, pause it and finalize the writers
    if let Some(stream) = stream_container.borrow_mut().as_mut() {
        stream.pause()?;
        finalize_writers_if_some(writer_handles).unwrap();
        println!("Restarting new recording...");
    } else {
        println!("Starting recording...");
    }

    // Make new writers
    let writers = smrec_config.writers()?;
    // Replace the old ones.
    writer_handles.lock().unwrap().replace(writers);

    // Errors when ctrl+c handler is already set. We ignore this error since we have no intention of a reset.
    let writer_handles_in_ctrlc = Arc::clone(writer_handles);
    let _ = ctrlc::try_set_handler(move || {
        // TODO: Necessary to drop stream?

        // TODO: Maybe inform user in unsuccessful operation?
        finalize_writers_if_some(&writer_handles_in_ctrlc).unwrap();

        // TODO: Better message, differentiate if the recording was stopped or interrupted.
        println!("\rRecording interrupted thus stopped.");
        std::process::exit(0);
    });

    // Create and start a new stream
    let new_stream = stream::build(
        device,
        smrec_config.supported_cpal_stream_config(),
        smrec_config.channels_to_record(),
        Arc::clone(writer_handles),
    )?;

    new_stream.play()?;
    println!("Recording started.");
    stream_container.borrow_mut().replace(new_stream);

    Ok(())
}

pub fn stop_recording(
    stream_container: &Rc<RefCell<Option<cpal::Stream>>>,
    writer_handles: &Arc<Mutex<Option<WriterHandles>>>,
) -> Result<()> {
    println!("Stopping recording...");

    if let Some(stream) = stream_container.borrow_mut().take() {
        stream.pause()?;
        finalize_writers_if_some(writer_handles)?;
        println!("Recording stopped.");
        return Ok(());
    }
    println!("There is no running recording to stop.");

    Ok(())
}

pub fn finalize_writers_if_some(writers: &Arc<Mutex<Option<WriterHandles>>>) -> Result<()> {
    let writers = writers.lock().unwrap().take();
    if let Some(writers) = writers {
        for writer in writers.iter() {
            if let Some(writer) = writer.lock().unwrap().take() {
                writer.finalize().unwrap();
            }
        }
    }
    Ok(())
}
